src/actions/pkg.cr

```diff
@@ -1,45 +1,155 @@
-module DppmRestApi::Actions::Pkg
-  extend self
-  # List built packages
-  relative_get nil do |context|
-    if context.current_user? && Actions.has_access? context, Access::Read
-      # TODO: list all built packages
-      next context
-    end
-    raise Unauthorized.new context
-  end
-  # Clean unused built packages
-  relative_delete "/clean" do |context|
-    if context.current_user? && Actions.has_access? context, Access::Delete
-      # TODO: delete all unused built packages
-      next context
+require "../ext/semantic_version"
+
+module DppmRestApi::Actions
+  module Pkg
+    extend self
+    ALL_PKGS = ""
+    ONE_PKG  = "/:id"
+    include RouteHelpers
+
+    def build_config_response(package, builder)
+      builder.field package.package do
+        builder.object do
+          package.each_config_key do |key|
+            builder.field key, package.get_config key
+          end
+        end
+      end
     end
-    raise Unauthorized.new context
-  end
-  # Query information about a given package
-  relative_get "/:id/query" do |context|
-    if context.current_user? && Actions.has_access? context, Access::Read
-      # TODO: Query information about the given package
-      next context
+
+    def clean_unused_packages(context : HTTP::Server::Context)
+      if result = DppmRestApi.prefix.clean_unused_packages(confirmation: false) { }
+        return {data: result} if result.any?
+        bug = InternalServerError.new context, "received empty set from Prefix#clean_unused_packages; please report this strange bug"
+        error = NoPkgsToClean.new context
+        error.cause = bug
+        raise error
+      end
+      raise NoPkgsToClean.new context
     end
-    raise Unauthorized.new context
-  end
-  # Delete a given package
-  relative_delete "/:id/delete" do |context|
-    if context.current_user? && Actions.has_access? context, Access::Delete
-      # TODO: Query information about the given package
-      next context
+
+    struct ListResponse
+      include JSON::Serializable
+      property version : SemanticVersion
+
+      def initialize(@version); end
+    end
+
+    # List built packages
+    relative_get nil do |context|
+      if Actions.has_access? context, Access::Read
+        pkgs = {} of String => ListResponse
+        DppmRestApi.prefix.each_pkg do |package|
+          if (found = pkgs[package.package]?) && found.version > package.semantic_version
+            # newer version of package already exists
```

Comment:
Why does it matter? We want to list **all** packages.
I think the best would be an array of hash of string => string.

Reply:
OK so the structure of the response is like so

```json
{ "data": {
  "dppm": {
    "version": "2019.4.26"
  }
} }
```

This leaves room for expanding the API to have further information such as 

```json
{ "data": {
  "dppm": {
    "version": "2019.4.26",
    "license": "ISC",
    "description": "Dedicated Platform Package Manager - an easy way to install and manage server applications."
  }
} }
```
I thought you wanted me to remove duplicates in this method but now that I think about it that doesn't really make sense. I'll get rid of that part.

Reply:
What duplicates? With the examples you posted, it can't have several `dppm` packages, which can happen when upgrading for example.

Reply:
I propose:
```json
"data":[
{
  "package": "dppm",
  "version": "1.2.3"
},
{
  "package": "dppm",
  "version": "1.2.7"
}
]
```
Of course it can also include description, license etc. as you said, depending of parameters.
It would be more or less a list of what's in `PkgFile`, plus the version.